crates/class-hash/src/lib.rs

```diff
@@ -138,9 +171,95 @@ pub mod from_parts {
+    // hash
+    prepare_json_contract_definition(&mut contract_definition)?;
+
+    let truncated_keccak = compute_cairo_hinted_class_hash(&contract_definition)?;
+
+    // what follows is defined over at the contract.cairo
+
+    const API_VERSION: Felt = Felt::ZERO;
+
+    let mut outer = HashChain::default();
+
+    // This wasn't in the docs, but similarly to contract_state hash, we start with
+    // this 0, so this will yield outer == H(0, 0); However, dissimilarly to
+    // contract_state hash, we do include the number of items in this
+    // class_hash.
+    outer.update(API_VERSION);
+
+    // It is important to process the different entrypoint hashchains in correct
+    // order. Each of the entrypoint lists gets updated into the `outer`
+    // hashchain.
+    //
+    // This implementation doesn't preparse the strings, which makes it a bit more
+    // noisy. Late parsing is made in an attempt to lean on the one big string
+    // allocation we've already got, but these three hash chains could be
+    // constructed at deserialization time.
+    [External, L1Handler, Constructor]
+        .iter()
+        .map(|key| {
+            contract_definition
+                .entry_points_by_type
+                .get(key)
+                .unwrap_or(&Vec::new())
+                .iter()
+                // flatten each entry point to get a list of (selector, offset, selector, offset,
+                // ...)
+                .flat_map(|x| [x.selector.0, x.offset.0].into_iter())
+                .fold(HashChain::default(), |mut hc, next| {
+                    hc.update(next);
+                    hc
+                })
+        })
+        .for_each(|x| outer.update(x.finalize()));
+
+    fn update_hash_chain(mut hc: HashChain, next: Result<Felt, Error>) -> Result<HashChain, Error> {
+        hc.update(next?);
+        Result::<_, Error>::Ok(hc)
+    }
+
+    let builtins = contract_definition
+        .program
+        .builtins
+        .iter()
+        .enumerate()
+        .map(|(i, s)| (i, s.as_bytes()))
+        .map(|(i, s)| {
+            Felt::from_be_slice(s).with_context(|| format!("Invalid builtin at index {i}"))
+        })
+        .try_fold(HashChain::default(), update_hash_chain)
+        .context("Failed to process contract_definition.program.builtins")?;
+
+    outer.update(builtins.finalize());
+
+    outer.update(truncated_keccak);
+
+    let bytecodes = contract_definition
+        .program
+        .data
+        .iter()
+        .enumerate()
+        .map(|(i, s)| {
+            Felt::from_hex_str(s).with_context(|| format!("Invalid bytecode at index {i}"))
+        })
+        .try_fold(HashChain::default(), update_hash_chain)
+        .context("Failed to process contract_definition.program.data")?;
+
+    outer.update(bytecodes.finalize());
+
+    Ok(ClassHash(outer.finalize()))
+}
+
+/// Prepare a Cairo contract definition to be compatible with the hinted class
+/// hash
```

Comment:
I'm not sure any user would know what a hinted class is.

Reply:
Ironically, this is exactly the method the guys at SNOS need 😂